src/client.rs

```diff
@@ -1280,6 +1280,91 @@ impl<T: Read + Write> Session<T> {
         .and_then(|lines| parse_id_seq(&lines, &mut self.unsolicited_responses_tx))
     }
 
+    /// The [`SETACL` command](https://datatracker.ietf.org/doc/html/rfc4314#section-3.1)
+    /// Modifies the ACLs on the given mailbox for the specified identifier.
+    /// Return [`Error::No`] if the logged in user does not have `a` rights on the mailbox.
```

Comment:
We should at the very least document that these methods aren't supported by all servers, and how users might check for server support. Alternatively, we could go one step further and do:

```rust
#[doc(alias = "RFC 4313")]
pub fn extension_acls(&mut self) -> Result<extensions::Acl<'_>> {
    if self.capabilities()?.has(Capability::TODO) {
        Ok(extensions::Acl(self))
    } else {
        Err("server does not support RFC 4314 ACL extensions")
    }
}
```

where all these methods are provided on the `extensions::Acl` type. That would also be a nice way to encapsulate other potential extensions in the future. What do you think?

Reply:
In fact, with GATs we could go even further and have

```rust
pub trait Extension {
    type Extended<'session>;
    fn extend<'session>(self, s: &'session mut Session) -> Result<Self::Extended<'session>>;
}

pub struct Rfc4314;

impl Extension for Rfc4314 {
    type Extended<'session> = extensions::Acl<'session>;
    fn extend<'session>(self, s: &'session mut Session) -> Result<Self::Extended<'session>> {
        if session.capabilities()?.has(Capability::TODO) {
            Ok(extensions::Acl(session))
        } else {
            Err("server does not support RFC 4314 ACL extensions")
        }
    }
}

impl Session {
    pub fn extension<E: Extension>(&mut self, ext: E) -> Result<E::Extended<'_>> {
        ext.extend(self)
    }
}
```

but given that GATs aren't stable I mostly just wanted to write that because it was interesting :)

Reply:
@jonhoo please look through the latest code changes and the two fixup commits that adjust the structure of things (the Cow'ifying).   Once we settle on those approaches then I'll work on this refactoring.

Reply:
I still think it'd be nice to do something like this, not just for this but also for the QRESYNC stuff that @bitfehler has been landing recently. We can do that in a separate PR though.

Reply:
Let's document that this (and the other ACL methods) only work for servers with the ACL-related capability listed though :+1:

Reply:
Leaving this unresolved to remember that we'll probably still want to make some change to how extensions are handled so that it's harder to misuse them and so that we don't pollute `Client` with tons and tons of methods.